Dockerfile

```diff
@@ -1,22 +1,43 @@
 FROM golang:1.22-alpine3.19 AS builder
 
-RUN apk add --no-cache make
+# Create a non-root user
+RUN addgroup -g 1001 -S iamgroup && \
+	adduser -u 1001 -S iamuser -G iamgroup
 
-RUN mkdir -p /go/src/github.com/status-im/telemetry
 WORKDIR /go/src/github.com/status-im/telemetry
-ADD . .
-RUN make build
 
-# Copy the binary to the second image
+# Copy go mod and sum files for caching
+COPY go.mod go.sum ./
+
+# Using mounts to cache dependencies
+RUN --mount=type=cache,target=/go/pkg/mod/ \
+	--mount=type=cache,target=/root/.cache/go-build \
+	go mod download -x
+
+COPY . .
+
+# Build the binary with static linking and stripped symbols for smaller size
+RUN GOOS=linux GOARCH=amd64 \
+	go build \
+	-ldflags "-s -w -extldflags -static" \
+	-o build/server cmd/server/main.go
```

Comment:
How is this better than just `make build`? And why should those flags not be used in the `Makefile`?

Reply:
`-s -w` flags are for debugging, so `-s` is to omit the symbol table, and `-w` is to omit DWARF tables, which is required for debugging, and `-extldflags -static` is for static linking. Now, it is not recommended omitting the debugging info while developing, so that is why I think these flags shouldn't be added in `make build`.

Reply:
The reason for omitting the `make build` is that we are using go alpine image, which don't come with `make` by default, and installing make will be an extra step to the build process. Also, using `go build` directly provides more transparency, so anyone maintaining the Dockerfile can exactly see what is happening in each stage without having to look into the Makefile.

Reply:
I see, fair enough.